core/src/main/java/org/wildfly/plugin/core/Deployment.java

```diff
@@ -0,0 +1,299 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.wildfly.plugin.core;
+
+import java.io.File;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Represents a deployment to be deployed or redeployed to a server.
+ *
+ * @author <a href="mailto:jperkins@redhat.com">James R. Perkins</a>
+ */
+@SuppressWarnings({"unused", "WeakerAccess"})
+public class Deployment implements DeploymentDescription {
+
+    private final Path pathContent;
+    private final InputStream inputStreamContent;
+    private final Set<String> serverGroups;
+    private String name;
+    private String runtimeName;
+    private boolean enabled;
+
+    /**
+     * Creates a new deployment for the file. If the file is a directory the content will be deployed exploded using
+     * the file system location.
+     *
+     * @param content the file containing the content
+     */
+    public Deployment(final File content) {
+        this(Assertions.requiresNotNullParameter(content, "content").toPath());
+    }
+
+    /**
+     * Creates a new deployment for the path. If the path is a directory the content will be deployed exploded using
+     * the file system location.
+     *
+     * @param content the path containing the content
+     */
+    public Deployment(final Path content) {
+        this.pathContent = Assertions.requiresNotNullParameter(content, "content");
+        inputStreamContent = null;
+        serverGroups = new LinkedHashSet<>();
+    }
+
+    /**
+     * Creates a new deployment for the input stream. The name is required for the deployment and cannot be {@code null}.
+     * If {@link #setName(String)} with a {@code null} argument is invoked when using this constructor an
+     * {@link IllegalArgumentException} will be thrown.
+     *
+     * @param content the input stream representing the content
+     * @param name    the name for the deployment
+     */
+    public Deployment(final InputStream content, final String name) {
+        this.pathContent = null;
+        inputStreamContent = Assertions.requiresNotNullParameter(content, "content");
+        serverGroups = new LinkedHashSet<>();
+        this.name = Assertions.requiresNotNullParameter(name, "name");
+    }
+
+    /**
+     * Creates a new deployment for the file. If the file is a directory the content will be deployed exploded using
+     * the file system location.
+     *
+     * @param content the file containing the content
+     *
+     * @return the deployment
+     */
+    public static Deployment of(final File content) {
```

Comment:
I mainly just had both to satisfy users. Some people prefer `new` some prefer static factories. I'll think about limiting the scope though.